nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -761,53 +715,41 @@ static void
 ngx_http_modsecurity_detection_thread_func(void *data, ngx_log_t *log)
 {
     // Executed in a separate thread
-    ngx_http_modsecurity_detection_thread_ctx_t *thread_ctx = data;
+    ngx_http_modsecurity_thread_ctx_t *thread_ctx = data;
+    ngx_http_modsecurity_ctx_t *mod_ctx = ngx_http_get_module_ctx(thread_ctx->r, ngx_http_modsecurity);
 
     // Processing request headers
-    ngx_int_t rc = ngx_http_modsecurity_status(modsecProcessRequestHeaders(thread_ctx->req));
+    ngx_int_t rc = ngx_http_modsecurity_status(modsecProcessRequestHeaders(mod_ctx->req));
     if (rc != NGX_DECLINED) {
         return;
     }
 
-    if (modsecContextState(thread_ctx->req) == MODSEC_DISABLED) {
+    if (modsecContextState(mod_ctx->req) == MODSEC_DISABLED) {
         return;
     }
 
     // The name of modsecProcessRequestBody is a bit misleading. This function call is needed even to just process GET args.
-    modsecProcessRequestBody(thread_ctx->req);
+    modsecProcessRequestBody(mod_ctx->req);
 }
 
-
 static void
 ngx_http_modsecurity_detection_thread_completion(ngx_event_t *ev)
 {
     // executed in nginx event loop after thread task is done, in order to pick up and continue processing request
-    ngx_http_modsecurity_detection_thread_ctx_t *ctx = ev->data;
-
-    request_rec *req = ctx->req;
-    conn_rec *connection = ctx->connection;
-
-    if (req != NULL) {
-        modsecFinishRequest(req);
-    }
-    if (connection != NULL) {
-        modsecFinishConnection(connection);
-    }
+    ngx_http_request_t *r = ev->data;
+    --r->main->blocked;
```

Comment:
I'm a little confused. It used to be `--` previously too. Now that we're doing `++` an extra time, should it be `-=2`?

Reply:
It is incremented once for a detection mode and once for a prevention mode. These modes have different completion handlers. We have been decrementing `blocking` for prevention mode for a long time now but never before used `count` and `blocking` for detection mode. Now we do, so this is why the line above has been added.